packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -136,112 +251,138 @@ export class Wallet {
-      throw new Error(`No messages for objective ${objectiveId}`);
-
-      // This is a sanity check to prevent us from losing messages
-    } else if (objectiveIds.length !== 1) {
-      throw new Error(`There are messages for multiple objectives ${objectiveIds}`);
-    }
-
-    return result.messagesByObjective[objectiveId];
+  private async handleEngineOutput(
+    output: MultipleChannelOutput | SingleChannelOutput
+  ): Promise<void> {
+    this.emitObjectiveEvents(output);
+    await this._messageService.send(getMessages(output));
+    await this._chainService.handleChainRequests(output.chainRequests);
   }
 
   public get messageService(): MessageServiceInterface {
     return this._messageService;
   }
 
+  private async createObjectiveDoneResult(
+    objective: WalletObjective
+  ): Promise<ObjectiveDoneResult> {
+    return new Promise<ObjectiveDoneResult>(resolve => {
+      this.on('ObjectiveTimedOut', o => {
+        if (o.objectiveId === objective.objectiveId) {
+          this._engine.logger.trace({objective: o}, 'Objective Timed out');
+
+          resolve({
+            type: 'ObjectiveTimedOutError',
+            objectiveId: o.objectiveId,
+            lastProgressMadeAt: o.progressLastMadeAt,
+          });
+        }
+      });
+      this.on('ObjectiveCompleted', o => {
+        if (o.objectiveId === objective.objectiveId) {
+          this._engine.logger.trace({objective: o}, 'Objective Suceeded');
+          resolve({type: 'Success', channelId: o.data.targetChannelId});
+        }
+      });
+    });
+  }
+
+  /**
+   * Registers any channels existing in the database with the chain service.
+   *
+   * @remarks
+   * Enables the chain service to alert the engine of of any blockchain events for existing channels.
+   *
+   * @returns A promise that resolves when the channels have been successfully registered.
+   */
+  private async registerExistingChannelsWithChainService(): Promise<void> {
+    const channelsToRegister = (await this._engine.store.getNonFinalizedChannels()).map(
+      ChannelState.toChannelResult
+    );
+    await this.registerChannels(channelsToRegister);
+  }
+
+  private async registerChannels(channelsToRegister: ChannelResult[]): Promise<void> {
+    const channelsWithAssetHolders = channelsToRegister.map(cr => ({
+      assetHolderAddresses: cr.allocations.map(a => makeAddress(a.assetHolderAddress)),
+      channelId: cr.channelId,
+    }));
+
+    for (const {channelId, assetHolderAddresses} of channelsWithAssetHolders) {
+      this._chainService.registerChannel(channelId, assetHolderAddresses, this._chainListener);
+    }
+  }
+
+  private createChainEventlistener<
+    K extends keyof ChainEventSubscriberInterface,
+    EH extends ChainEventSubscriberInterface[K]
+  >(eventName: K, storeUpdater: EH) {
+    return async (
+      event: HoldingUpdatedArg &
+        AssetOutcomeUpdatedArg &
+        ChannelFinalizedArg &
+        ChallengeRegisteredArg
```

Comment:
This is a strange type -- I'd expect it to be
```suggestion
      event: HoldingUpdatedArg ||
        AssetOutcomeUpdatedArg ||
        ChannelFinalizedArg ||
        ChallengeRegisteredArg
```

What's the point of the intersection type?